[IMP] DV, CF: Enforce range on CF & DV

fdamhaut opened the pull request:
## Description:

Enforce that the ranges on CF & DV belong to the same sheet they do.

Task: [4072640](https://www.odoo.com/odoo/2328/tasks/4072640)

## review checklist

- [ ] feature is organized in plugin, or UI components
- [ ] support of duplicate sheet (deep copy)
- [ ] in model/core: ranges are Range object, and can be adapted (adaptRanges)
- [ ] in model/UI: ranges are strings (to show the user)
- [ ] undo-able commands (uses this.history.update)
- [ ] multiuser-able commands (has inverse commands and transformations where needed)
- [ ] new/updated/removed commands are documented
- [ ] exportable in excel
- [ ] translations (\_t("qmsdf %s", abc))
- [ ] unit tested
- [ ] clean commented code
- [ ] track breaking changes
- [ ] doc is rebuild (npm run doc)
- [ ] status is correct in Odoo

robodoo commented:
[![Pull request status dashboard](https://mergebot.odoo.com/odoo/o-spreadsheet/pull/5679.png)](https://mergebot.odoo.com/odoo/o-spreadsheet/pull/5679)

hokolomopo commented on src/components/side_panel/conditional_formatting/conditional_formatting.ts:
```diff
@@ -46,18 +46,18 @@ export class ConditionalFormattingPanel extends Component<Props, SpreadsheetChil
+      if (this.state.mode == "list") {
+        this.activeSheetId = this.env.model.getters.getActiveSheetId();
+      }
+      if (nextProps.selection !== this.props.selection) {
+        const rules = this.env.model.getters.getRulesSelection(
+          this.activeSheetId,
+          nextProps.selection || []
+        );
+        const cf =
+          rules.length === 1 ? this.conditionalFormats.find((c) => c.id === rules[0]) : undefined;
```
why change this ?

hokolomopo commented on src/components/side_panel/conditional_formatting/conditional_formatting.ts:
```diff
@@ -46,18 +46,18 @@ export class ConditionalFormattingPanel extends Component<Props, SpreadsheetChil
-            this.editConditionalFormat(cf);
-          }
+      if (this.state.mode == "list") {
+        this.activeSheetId = this.env.model.getters.getActiveSheetId();
```
`this.activeSheetId` is useless no ? it's only used when we're in lsit mode, and this make sure that it's always equal to `getters.getActiveSheetId`. Just use `getters.getActiveSheetId` directly.

fdamhaut replied:
It's also used as the defaultSheetId for transforming stringXC to range on save (not in list mode, so still to the value of the rule "main" sheet.

hokolomopo commented on src/components/translations_terms.ts:
```diff
@@ -28,6 +28,7 @@ export const CfTerms = {
     [CommandResult.ValueCellIsInvalidFormula]: _t(
       "At least one of the provided values is an invalid formula"
     ),
+    [CommandResult.TargetOutOfSheet]: _t("The range cannot target out of the current sheet."),
```
Message isn't very clear IMO. Since we can leave the side panel open while changing sheet, talking about current sheet is dubious. "The range" when there is no clear way to know about which range we are talking about is also not very clear.

"All the ranges should be in the same sheet as the conditional format" maybe ?

hokolomopo replied:
Still valid, the message 
![image](https://github.com/user-attachments/assets/42ad2d50-1ea7-45e3-be70-994e6e5439cc)

does not seems very english/clear. Maybe it's just me. And we should probably put the range between quotes.

hokolomopo reviewed (commented):
pouet pouet :trumpet:

hokolomopo commented on src/components/side_panel/conditional_formatting/cf_editor/cf_editor.ts:
```diff
@@ -198,6 +200,7 @@ export class ConditionalFormattingEditor extends Component<Props, SpreadsheetChi
   setup() {
     this.state = useState<State>({
       errors: [],
+      sheetId: this.env.model.getters.getActiveSheetId(),
```
not addressed

hokolomopo commented on src/components/side_panel/conditional_formatting/cf_editor/cf_editor.ts:
```diff
@@ -225,7 +224,33 @@ export class ConditionalFormattingEditor extends Component<Props, SpreadsheetChi
+  get firstInvalidRangeString(): string | undefined {
+    const sheetId = this.sheetId;
+    return this.state.ranges.find((xc) => {
+      const range = this.env.model.getters.getRangeDataFromXc(sheetId, xc);
```
```suggestion
      const range = this.env.model.getters.getRangeFromSheetXc(sheetId, xc);
```
don't use RangeData when you do'nt have to. They are a bit ugly, and only meant for commands. We *shouldn't* want to access range._sheetId, it's "private"

hokolomopo commented on src/components/translations_terms.ts:
```diff
@@ -5,30 +5,30 @@ import { CommandResult } from "../types/index";
+    [CommandResult.MaxNaN]: () => _t("The maxpoint must be a number"),
+    [CommandResult.ValueUpperInflectionNaN]: () => _t("The first value must be a number"),
+    [CommandResult.ValueLowerInflectionNaN]: () => _t("The second value must be a number"),
+    [CommandResult.MinBiggerThanMax]: () => _t("Minimum must be smaller then Maximum"),
+    [CommandResult.MinBiggerThanMid]: () => _t("Minimum must be smaller then Midpoint"),
+    [CommandResult.MidBiggerThanMax]: () => _t("Midpoint must be smaller then Maximum"),
```
while you're changing this: you could fix the message. "than" rather than "then". And remove random uppercasing of words

hokolomopo commented on src/plugins/core/conditional_format.ts:
```diff
@@ -144,6 +144,8 @@ export class ConditionalFormatPlugin
       case "ADD_CONDITIONAL_FORMAT":
         return this.checkValidations(
           cmd,
+          this.checkEmptyRange,
+          this.checkValidRange,
           this.checkCFRule,
           this.checkEmptyRange,
```
this.checkEmptyRange is there twice

hokolomopo commented on src/plugins/core/conditional_format.ts:
```diff
@@ -389,6 +391,18 @@ export class ConditionalFormatPlugin
+    if (ranges.some((range) => range.sheetId != cmd.sheetId)) {
+      return CommandResult.TargetOutOfSheet;
+    }
+    const stringRanges = ranges.map((range) => this.getters.getRangeString(range, cmd.sheetId));
+    if (stringRanges.some((xc) => !this.getters.isRangeValid(xc))) {
+      return CommandResult.InvalidRange;
+    }
```
probably wrong way to do that/useless. We're coming from a RangeData, so either it's a normal rangeData and the XC will always be valid, or it's really malformed and the conversion to range/string will throw and we'll never return `CommandResult.InvalidRange`.

And we already test for invalid zones inside the sheet core plugin.

hokolomopo replied:
Oh you copied what was already in the data validation. Well either i'm wrong and there is cases where you return `CommandResult.InvalidRange` here, or you can remove it from both.

fdamhaut replied:
I didn't look at it in details but your comment makes sense.

hokolomopo commented on tests/conditional_formatting/conditional_formatting_panel_component.test.ts:
```diff
@@ -427,8 +427,27 @@ describe("UI of conditional formats", () => {
+      setInputValueAndTrigger(selectors.ruleEditor.range, "Sheet2!A1");
+
+      const dispatch = spyModelDispatch(model);
+      //  click save
```
nitpick: comment looks useless

hokolomopo commented on tests/data_validation/data_validation_generics_side_panel_component.test.ts:
```diff
@@ -152,7 +152,26 @@ describe("data validation sidePanel component", () => {
+    createSheet(model, { sheetId: "ID", name: "Sheet2", activate: false });
+    await simulateClick(".o-dv-add");
+    await nextTick();
+    await changeCriterionType("dateIs");
```
useless no ?

hokolomopo commented on tests/data_validation/data_validation_generics_side_panel_component.test.ts:
```diff
@@ -18,7 +18,7 @@ mockGetBoundingClientRect({
 });
 
-export async function mountDataValidationPanel(model?: Model) {
+async function mountDataValidationPanel(model?: Model) {
```
remove this and use the new helper

fdamhaut replied:
What is 'the new helper' ?

hokolomopo replied:
Oh you duplicated the helper in both test files, I though you defined a new helper that you could import everywhere. 

Maybe you should define it somewhere so you can import it from both files ? But probably we don't care, it's only like 4 lines.

hokolomopo reviewed (commented):
:+1:

hokolomopo commented on src/components/helpers/formulas.ts:
```diff
@@ -0,0 +1,23 @@
+import { rangeTokenize } from "../../formulas";
+import { Getters } from "../../types/getters";
+
+export function adaptFormulaToSheet(
```
nitpick: maybe could be a getter in range ?

fdamhaut replied:
There are others functions that affect formulas coming to this file (in others PRs). I feel like range is already big enough and doesn't need more bloating.

hokolomopo commented on src/components/selection_input/selection_input.ts:
```diff
@@ -83,6 +85,10 @@ export class SelectionInput extends Component<Props, SpreadsheetChildEnv> {
     colors: { type: Array, optional: true, default: [] },
   };
+  static defaultProps = {
+    readonly: false,
```
nitpick: it's a bit useless , undefined is already falsy. And if you have a default props, the props should be typed as non-optional in `Props` (because it will always have a value in practice)

hokolomopo commented on src/components/selection_input/selection_input.ts:
```diff
@@ -97,19 +103,19 @@ export class SelectionInput extends Component<Props, SpreadsheetChildEnv> {
 
   get canAddRange(): boolean {
-    return !this.props.hasSingleRange;
+    return !this.props.hasSingleRange && !this.props.readonly;
```
Some issues with the readonly style:

- why does the reset/add range button disappear, but the confirm button stays here and is disabled ? We should stay consistent.
- the "trash" button should definitely not be there in readonly
- the selection input should probably lose its colors in readonly. Maybe cancel/confirm the edition to lose focus so we bother lose the colors and the buttons ?
- probably should change the border color rather than the border size when disabling the inputs, otherwise they all jump one pixel up

hokolomopo commented on src/components/selection_input/selection_input.xml:
```diff
@@ -15,14 +15,15 @@
           <t t-call="o-spreadsheet-Icon.SHORT_THIN_DRAG_HANDLE"/>
         </span>
-        <div class="position-relative w-100">
+        <div class="position-relative w-100" t-att-disabled="props.readonly">
```
is this disabled useful ?

fdamhaut replied:
not anymore

hokolomopo commented on src/components/helpers/formulas.ts:
```diff
@@ -0,0 +1,23 @@
+import { UID } from "../..";
```
see with mehdi because it's probably related to https://www.odoo.com/odoo/2328/tasks/4589387, but the thrash icon does not work when you enter a range from another sheet

fdamhaut replied:
It's not related to range from other sheet, in master it simply doesn't work at all :/.
I'll see with mehdi if he already has a fix

hokolomopo commented on src/components/side_panel/conditional_formatting/cf_editor/cf_editor.ts:
```diff
@@ -570,14 +605,18 @@ export class ConditionalFormattingEditor extends Component<Props, SpreadsheetChi
     return {
       onConfirm: (str: string) => {
-        this.state.rules.cellIs.values[valueIndex] = str;
+        this.state.rules.cellIs.values[valueIndex] = adaptFormulaToSheet(
+          this.env.model.getters,
+          str,
+          this.props.sheetId
+        );
```
what does this do ? When testing it never changes anything ? And I don't get why we could need to modify the ranges of the formula ?

fdamhaut replied:
If you start editing while one the other sheet, the composer use the current sheet as the 'defaultRangeSheetId' (Meaning A1 = s2!A1) which should be prefixed when saving the string

hokolomopo commented on src/components/side_panel/conditional_formatting/cf_editor/cf_editor.ts:
```diff
@@ -365,12 +395,17 @@ export class ConditionalFormattingEditor extends Component<Props, SpreadsheetChi
-    return this.state.errors.includes(CommandResult.SecondArgMissing);
+    return (
+      this.state.errors.includes(CommandResult.SecondArgMissing) ||
+      (this.state.errors.includes(CommandResult.ValueCellIsInvalidFormula) &&
+        this.invalidFormulas.includes(this.state.rules.cellIs.values[1]))
```
testing both sounds redundant

hokolomopo commented on src/components/side_panel/data_validation/dv_editor/dv_editor.ts:
```diff
@@ -134,6 +144,24 @@ export class DataValidationEditor extends Component<Props, SpreadsheetChildEnv>
+    }
+  }
+
+  get firstInvalidRangeString(): string | undefined {
```
why stop at the first ? it doesn't cost anything to return them all and join them in the error string no ?

hokolomopo commented on src/components/translations_terms.ts:
```diff
@@ -5,30 +5,36 @@ import { CommandResult } from "../types/index";
+    [CommandResult.TargetOutOfSheet]: (sheetName: string, ranges: string[]) =>
+      _t(
+        "Ranges '%s' should target the sheet on which the conditional format is defined (%s)",
+        ranges.join("', '"),
```
joint with `"` instead. `'` is used for sheet names with spaces (eg. `'Split Panes'!A1` )

hokolomopo commented on tests/conditional_formatting/conditional_formatting_panel_component.test.ts:
```diff
@@ -427,8 +427,38 @@ describe("UI of conditional formats", () => {
+      model.dispatch("ACTIVATE_NEXT_SHEET");
+      await nextTick();
+
+      const range = document.querySelector(selectors.ruleEditor.range) as HTMLInputElement;
```
```suggestion
      const range = fixture.querySelector<HTMLInputElement>(selectors.ruleEditor.range);
```
nitpick

hokolomopo commented on tests/data_validation/data_validation_generics_side_panel_component.test.ts:
```diff
@@ -152,7 +152,25 @@ describe("data validation sidePanel component", () => {
+    expect(errorMessageEl?.textContent).toContain("The range A1:HOLA is invalid.");
+  });
+
+  test("Out of sheet range", async () => {
```
mayeb also a test ·`cannot edit a DV range on another sheet`

hokolomopo commented on tests/data_validation/data_validation_list_component.test.ts:
```diff
@@ -36,6 +37,13 @@ beforeEach(async () => {
+
 describe("Edit criterion in side panel", () => {
   describe("Value in list", () => {
     beforeEach(async () => {
```
missing a test for the new props readonly of the selection input

hokolomopo reviewed (commented):
Some issues with the style of the disabled selection input. And I'm probably dumb, but I don't get what `adaptFormulaToSheet` does in practice/why it's needed ?  

Mostly nitpicks otherwise :)

hokolomopo commented on src/components/side_panel/data_validation/dv_editor/dv_editor.ts:
```diff
@@ -134,6 +144,25 @@ export class DataValidationEditor extends Component<Props, SpreadsheetChildEnv>
+  errorMessage(reason: CancelledReason): string {
+    switch (reason) {
+      case CommandResult.TargetOutOfSheet:
+        return DVTerms.Errors[reason](this.props.sheetId, this.invalidRangeString);
```
sheetName, not sheetId

hokolomopo commented on src/components/spreadsheet/spreadsheet.ts:
```diff
@@ -190,7 +190,14 @@ css/* scss */ `
+      }
+
+      &:disabled {
+        border-width: 0 0 0 0;
```
not needed anymore. You overwrite it back in the selection input css

hokolomopo commented on src/components/spreadsheet/spreadsheet.ts:
```diff
@@ -190,7 +190,14 @@ css/* scss */ `
+
+      &:disabled {
+        border-width: 0 0 0 0;
+        cursor: not-allowed;
```
Either put all the `disabled` css in the selection input css, or all the css here, I don't think there's a point of having half the css in each place

hokolomopo commented on src/components/spreadsheet/spreadsheet.ts:
```diff
@@ -190,7 +190,14 @@ css/* scss */ `
 
     .o-input[type="text"] {
-      border-width: 0 0 1px 0;
+      &:not([disabled]) {
```
same here, not needed anymore

hokolomopo reviewed (commented):
some small issues,  otherwise LGTM

fdamhaut committed:
[REF] composer: make cancelEdition modulary

Task-id: 4072640

fdamhaut committed:
[IMP] standalone compose: avoid sheet activation on blur

Task-id: 4072640

fdamhaut committed:
[IMP] DV, CF: Enforce range on CF & DV

- Allow to switch sheet while editing CF to set formulas
- Enforce that the target range of CF & DV belong to the same
sheet than them
- While on another sheet, range edition is disabled.

Task-id: 4072640

fdamhaut committed:
[IMP] ranges: rename applyRangeChange"CHANGE" to "RENAME"

For clarity sake
